dhall/src/Dhall/Lint.hs

```diff
@@ -42,15 +47,20 @@ import qualified Lens.Family
     * fixes @let a = x ≡ y@ to be @let a = assert : x ≡ y@
     * consolidates nested @let@ bindings to use a multiple-@let@ binding with 'removeLetInLet'
     * fixes paths of the form @.\/..\/foo@ to @..\/foo@
+    * Replaces deprecated @Optional\/fold@ and @Optional\/build@ built-ins
 -}
 lint :: Expr s Import -> Expr s Import
-lint = Dhall.Optics.rewriteOf subExpressions rewrite
+lint =  Dhall.Optics.rewriteOf subExpressions lowerPriorityRewrite
+    .   Dhall.Optics.rewriteOf subExpressions higherPriorityRewrite
   where
-    rewrite e =
-            fixAssert            e
-        <|> removeUnusedBindings e
-        <|> fixParentPath        e
-        <|> removeLetInLet       e
+    lowerPriorityRewrite e =
+            fixAssert                e
+        <|> removeUnusedBindings     e
+        <|> fixParentPath            e
+        <|> removeLetInLet           e
+        <|> replaceOptionalBuildFold e
+
+    higherPriorityRewrite = replaceSaturatedOptionalFold
```

Comment:
Ah, now I understand why my request was tricky to implement! :) Cheers!

Reply:
Yeah, the rewrite is depth-first, so the `Optional/fold` gets replaced before the rewrite has a chance to match on the larger saturated `Optional/fold` expression